DESIGN.md

```diff
@@ -1,31 +1,379 @@
+- Built-in into rabbitmq-server code
+- Home grown argument parsing
+- All commands in the same module (function clauses)
+- Started with `erl` with parameters
+- Too many commands
 
+All this made it hard to maintain and extend the tool.
 
-## Boundaries
-* The main module is responsible for the flow of the program. Anything that happens on any given execution of `rabbitmqctl` belongs in the main function. Currently its responsibilities are:
-	* parsing input (using `OptionParser`)
-	* auto-filling some default options
-	* execute commands using `eval_string`
-	* return specific exit codes based on output from the commands
+New CLI features to fix that:
 
-* The individual modules are responsible for any behavior unique to the command:
-	* Determining via pattern-matching whether the user-defined arguments and options are valid
-	* Interacting with the RabbitMQ server via RPC calls
-	* Returning particular outputs based on results from the server
-	* Printing any output associated with the command
-	* Providing any usage information via a `usage/0` method.
+- New repo on github.
+- Elixir language (with OptionParser)
+- A command per module
+- Extensible output formatting
+- Single executable escript file
+- Additional command scopes to organize commands in several tools (e.g. rabbitmq-plugins, rabbitmq-diagnostics)
 
-## Things The CLI Is Not Responsible For:
+## Architecture approach:
 
-* *Server behavior*: The CLI passes information to the server and triggers actions within it via RPC calls. It's up to the server to handle those calls properly. If the CLI passes valid credentials to RabbitMQ and RabbitMQ returns `{:refused, ...} `, the CLI has still functioned correctly.
+The main difference from the old CLI tool is command and output extensibility.
+
+Each command is defined in it's own module, implementing command behaviour. (see [Command behaviour](#command-behaviour))
+
+Output is processed with formatters and printers, which formats command output and renders it to somewhere. (see [Output formatting](#output-formatting))
+
+CLI core consists of several modules implementing command execution process:
+
+ - `RabbitMQCtl` - escript entry point. Command execution logic.
+ - `CommandModules` - responsible for command modules location
+ - `Helpers` - various functions to use in commands and main module
+ - `Config` - enviroment variable/argument values merge
+ - `Output` - output formatting logic
+ - `Parser` - OptionParser configuration
+
+### Command execution process:
+
+1. Arguments parsing
+
+ Command line arguments are parsed with OptionParser [http://elixir-lang.org/docs/stable/elixir/OptionParser.html]
+ Parser return a list of unnamed arguments and a map of options (named arguemtns)
```

Comment:
Parser return a list of unnamed arguments and a map of options (named arguemtns) => Parser returns a list of unnamed arguments and a map of options (named arguments).